### PR TITLE
OpenSwaps list per tenor tracking

### PR DESCRIPTION
**Pull request by pete-ipor**

### CODE REVIEW

File: contracts/interfaces/IAmmStorage.sol [hunk 1]
```diff
@@ -19,6 +20,8 @@ interface IAmmStorage {
     /// @return last swap ID, integer
     function getLastSwapId() external view returns (uint256);
 
+    function getLastOpenSwap(AmmTypes.SwapDuration duration, uint256 direction) external view returns (AmmInternalTypes.OpenSwapItem memory);
```
Comment (amydlarz) [inline contracts/interfaces/IAmmStorage.sol:23]: Would we like to add this to Lens?

**Review by amydlarz — APPROVED** [review]

File: contracts/amm/AmmStorage.sol [hunk 2]
```diff
@@ -57,8 +63,20 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
         return _lastSwapId;
     }
 
+    function getLastOpenSwap(AmmTypes.SwapDuration duration, uint256 direction)
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:66]: getLastOpenedSwap

File: contracts/amm/AmmStorage.sol [hunk 3]
```diff
@@ -780,6 +795,78 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
         );
     }
 
+    function _addNewOpenSwapPayFixed(
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:798]: I propose _updateLastOpenedSwapWhenOpenPayFixed

File: contracts/amm/AmmStorage.sol [hunk 4]
```diff
@@ -780,6 +795,78 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
...
+            AmmInternalTypes.OpenSwapItem memory swapNext = _openSwapsPayFixed[duration].swaps[swap.nextSwapId];
+            swapPrev.nextSwapId = swapNext.swapId;
+            swapNext.previousSwapId = swapPrev.swapId;
+            _openSwapsPayFixed[duration].swaps[swap.previousSwapId] = swapPrev;
+            _openSwapsPayFixed[duration].swaps[swap.nextSwapId] = swapNext;
+            delete _openSwapsPayFixed[duration].swaps[swapId.toUint32()];
+        }
+    }
+
+    function _addNewOpenSwapReceiveFixed(
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:834]: _updateLastOpenedSwapWhenOpenReceiveFixed

File: contracts/amm/AmmStorage.sol [hunk 5]
```diff
@@ -780,6 +795,78 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
...
+            swapId.toUint32(),
+            0,
+            headSwapId,
+            openTimestamp.toUint32()
+        );
+        _openSwapsPayFixed[duration].headSwapId = swapId.toUint32();
+        _openSwapsPayFixed[duration].swaps[headSwapId].nextSwapId = swapId.toUint32();
+    }
+
+    function _removeOpenSwapPayFixed(AmmTypes.SwapDuration duration, uint256 swapId) internal {
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:814]: _updateLastOpenedSwapWhenClosePayFixed

File: contracts/amm/AmmStorage.sol [hunk 6]
```diff
@@ -32,14 +35,17 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
     StorageInternalTypes.IporSwapContainer internal _swapsReceiveFixed;
 
     mapping(address => uint128) private _liquidityPoolAccountContribution;
+    mapping(AmmTypes.SwapDuration => AmmInternalTypes.OpenSwapList) private _openSwapsPayFixed;
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:38]: _openedSwapsPayFixed

File: contracts/amm/AmmStorage.sol [hunk 7]
```diff
@@ -780,6 +795,78 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
         );
     }
 
+    function _updateLastOpenedSwapWhenOpenPayFixed(
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:798]: I think word "last" is redundant

File: contracts/amm/AmmStorage.sol [hunk 8]
```diff
@@ -780,6 +795,78 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
...
+            swapId.toUint32(),
+            0,
+            headSwapId,
+            openTimestamp.toUint32()
+        );
+        _openSwapsPayFixed[duration].headSwapId = swapId.toUint32();
+        _openSwapsPayFixed[duration].swaps[headSwapId].nextSwapId = swapId.toUint32();
+    }
+
+    function _updateLastOpenedSwapWhenClosePayFixed(AmmTypes.SwapDuration duration, uint256 swapId) internal {
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:814]: I think word "last" is redundant

File: contracts/amm/AmmStorage.sol [hunk 9]
```diff
@@ -780,6 +795,78 @@ contract AmmStorage is Initializable, PausableUpgradeable, UUPSUpgradeable, Ipor
...
+        uint32 headSwapId = _openSwapsPayFixed[duration].headSwapId;
+        AmmInternalTypes.OpenSwapItem memory swap = _openSwapsPayFixed[duration].swaps[swapId.toUint32()];
+        if (swapId.toUint32() == headSwapId) {
+            AmmInternalTypes.OpenSwapItem memory swapPrev = _openSwapsPayFixed[duration].swaps[swap.previousSwapId];
+            swapPrev.nextSwapId = 0;
+            _openSwapsPayFixed[duration].headSwapId = swapPrev.swapId;
+            _openSwapsPayFixed[duration].swaps[swapPrev.swapId] = swapPrev;
+            delete _openSwapsPayFixed[duration].swaps[swapId.toUint32()];
+        } else {
+            AmmInternalTypes.OpenSwapItem memory swapPrev = _openSwapsPayFixed[duration].swaps[swap.previousSwapId];
```
Comment (mario-ipor) [inline contracts/amm/AmmStorage.sol:824]: What if swapId is old one, from version 1 of protocol?

Comment (pete-ipor) [reply contracts/amm/AmmStorage.sol:824]: i added check 
```
if(swap.openSwapTimestamp == 0) {
            return;
        }
```

so we will ignore old one

**Review by mario-ipor — APPROVED** [review]

**Pull request merged**